examples/seismic/viscoacoustic/operators.py

```diff
@@ -130,14 +287,157 @@ def deng_mcmechan(model, geometry, v, p, **kwargs):
 
 
-def ForwardOperator(model, geometry, space_order=4, kernel='blanch_symes', save=False,
-                    **kwargs):
+def deng_2nd_order(model, geometry, v, p, **kwargs):
+    """
+    Implementation of the 2nd order viscoacoustic wave-equation
+    from Deng and McMechan (2007).
+
+    https://library.seg.org/doi/pdf/10.1190/1.2714334
+
+    Parameters
+    ----------
+    v : VectorTimeFunction
+        Particle velocity.
+    p : TimeFunction
+        Pressure field.
+    """
+    forward = kwargs.get('forward', True)
+    space_order = p.space_order
+
+    s = model.grid.stepping_dim.spacing
+    f0 = geometry._f0
+    vp = model.vp
+    b = model.b
+    qp = model.qp
+    damp = model.damp
+
+    # Angular frequency
+    w0 = 2. * np.pi * f0
+
+    # Density
+    rho = 1. / b
+
+    bm = rho * (vp * vp)
+
+    if forward:
+
+        # Auxiliary functions
+        w = VectorTimeFunction(name="w", grid=model.grid, time_order=2,
+                               space_order=space_order)
+        h = TimeFunction(name="h", grid=model.grid, time_order=2, space_order=space_order,
+                         staggered=NODE)
+
+        u_h = Eq(h, (p - p.backward) / s)
+
+        u_v = Eq(v, b * grad(p))
+
+        u_w = Eq(w, b * grad(h))
+
+        pde_p = 2. * p - damp * p.backward + s * s * bm * div(v) - \
+            s * s * (w0 / qp) * h
+        u_p = Eq(p.forward, damp * pde_p)
+
+        return [u_h, u_v, u_w, u_p]
+
+    else:
+
+        # Auxiliary functions
+        h = TimeFunction(name="h", grid=model.grid, time_order=2, space_order=space_order,
+                         staggered=NODE)
+        g = TimeFunction(name="g", grid=model.grid, time_order=2, space_order=space_order,
+                         staggered=NODE)
+
+        u_h = Eq(h, bm * p)
+
+        u_v = Eq(v, b * grad(h))
+
+        u_g = Eq(g, (p.forward - p) / s)
+
+        pde_p = 2. * p - damp * p.forward + s * s * div(v) + \
+            s * s * (w0 / qp) * g
+        u_p = Eq(p.backward, damp * pde_p)
+
+        return [u_h, u_v, u_g, u_p]
+
+
+def sls(model, geometry, v, p, forward=True, **kwargs):
+    """
+    Implementation of the 1st order viscoacoustic wave-equation
+    from Blanch and Symes (1995) / Dutta and Schuster (2014) and
+    Implementation of the 2nd order viscoacoustic wave-equation from Bai (2014).
+
+    https://library.seg.org/doi/pdf/10.1190/1.1822695
+    https://library.seg.org/doi/pdf/10.1190/geo2013-0414.1
+    https://library.seg.org/doi/10.1190/geo2013-0030.1
+
+    Parameters
+    ----------
+    v : VectorTimeFunction
+        Particle velocity.
+    p : TimeFunction
+        Pressure field.
+    """
+    time_order = p.time_order
+
+    eq_stencil = stencils[('sls', time_order)]
+    eqn = eq_stencil(model, geometry, v, p, forward=forward, save=kwargs.get('save'))
+
+    return eqn
+
+
+def ren(model, geometry, v, p, forward=True, **kwargs):
+    """
+    Implementation of the 1st order viscoacoustic wave-equation and 2nd order
```

Comment:
Maybe: Implementation of the 1st and 2nd order viscoacoustic wave-equation from Ren et al. (2014).